### PR TITLE
Add back ZStream.effectAsync to JS

### LINKED ISSUE #3490: ZStream.effectAsync* methods are unsupported on Scala.js
These methods need to block with `runtime.unsafeRun`, so they're unsupported on Scala.js.

It would be interesting to explore a Scala.js implementation that does not block through the use of native promises.

### PR DESCRIPTION
**Pull request by luis3m**
Closes #3490

A follow-up of #3793 with the other effectAsync variants as requested

### CODE REVIEW

File: streams/js/src/main/scala/zio/stream/platform.scala [hunk 1]
```diff
@@ -2,12 +2,132 @@ package zio.stream
...
+    )
+
+  /**
+   * Creates a stream from an asynchronous callback that can be called multiple times.
+   * The registration of the callback returns either a canceler or synchronously returns a stream.
+   * The optionality of the error type `E` can be used to signal the end of the stream, by
+   * setting it to `None`.
+   */
+  def effectAsyncInterrupt[R, E, A](
+    register: (ZIO[R, Option[E], Chunk[A]] => Future[Unit]) => Either[Canceler[R], ZStream[R, E, A]],
```
Comment (iravid) [inline streams/js/src/main/scala/zio/stream/platform.scala:37]:
```suggestion
    register: (ZIO[R, Option[E], Chunk[A]] => Future[Boolean]) => Either[Canceler[R], ZStream[R, E, A]],
```

File: streams/js/src/main/scala/zio/stream/platform.scala [hunk 2]
```diff
@@ -2,12 +2,132 @@ package zio.stream
...
+    outputBuffer: Int = 16
+  ): ZStream[R, E, A] =
+    ZStream {
+      for {
+        output  <- Queue.bounded[stream.Take[E, A]](outputBuffer).toManaged(_.shutdown)
+        runtime <- ZIO.runtime[R].toManaged_
+        eitherStream <- ZManaged.effectTotal {
+                          register(k =>
+                            try {
+                              runtime.unsafeRunToFuture(stream.Take.fromPull(k).flatMap(output.offer).unit)
```
Comment (iravid) [inline streams/js/src/main/scala/zio/stream/platform.scala:47]:
```suggestion
                              runtime.unsafeRunToFuture(stream.Take.fromPull(k).flatMap(output.offer))
```

File: streams/js/src/main/scala/zio/stream/platform.scala [hunk 3]
```diff
@@ -2,12 +2,132 @@ package zio.stream
...
+                            } catch {
+                              case FiberFailure(c) if c.interrupted =>
+                                Future.successful(())
```
Comment (iravid) [inline streams/js/src/main/scala/zio/stream/platform.scala:50]:
```suggestion
                                Future.successful(false)
```

File: streams/js/src/main/scala/zio/stream/platform.scala [hunk 4]
```diff
@@ -2,12 +2,132 @@ package zio.stream
...
+      } yield pull
+    }
+
+  /**
+   * Creates a stream from an asynchronous callback that can be called multiple times
+   * The registration of the callback itself returns an effect. The optionality of the
+   * error type `E` can be used to signal the end of the stream, by setting it to `None`.
+   */
+  def effectAsyncM[R, E, A](
+    register: (ZIO[R, Option[E], Chunk[A]] => Future[Unit]) => ZIO[R, E, Any],
```
Comment (iravid) [inline streams/js/src/main/scala/zio/stream/platform.scala:74]:
```suggestion
    register: (ZIO[R, Option[E], Chunk[A]] => Future[Boolean]) => ZIO[R, E, Any],
```

File: streams/js/src/main/scala/zio/stream/platform.scala [hunk 5]
```diff
@@ -2,12 +2,132 @@ package zio.stream
...
+    outputBuffer: Int = 16
+  ): ZStream[R, E, A] =
+    managed {
+      for {
+        output  <- Queue.bounded[stream.Take[E, A]](outputBuffer).toManaged(_.shutdown)
+        runtime <- ZIO.runtime[R].toManaged_
+        _ <- register { k =>
+               try {
+                 runtime.unsafeRunToFuture(stream.Take.fromPull(k).flatMap(output.offer).unit)
```
Comment (iravid) [inline streams/js/src/main/scala/zio/stream/platform.scala:83]:
```suggestion
                 runtime.unsafeRunToFuture(stream.Take.fromPull(k).flatMap(output.offer))
```

File: streams/js/src/main/scala/zio/stream/platform.scala [hunk 6]
```diff
@@ -2,12 +2,132 @@ package zio.stream
...
+               } catch {
+                 case FiberFailure(c) if c.interrupted =>
+                   Future.successful(())
```
Comment (iravid) [inline streams/js/src/main/scala/zio/stream/platform.scala:86]:
```suggestion
                   Future.successful(false)
```

File: streams/js/src/main/scala/zio/stream/platform.scala [hunk 7]
```diff
@@ -2,12 +2,132 @@ package zio.stream
...
+  def effectAsyncMaybe[R, E, A](
+    register: (ZIO[R, Option[E], Chunk[A]] => Future[Boolean]) => Option[ZStream[R, E, A]],
+    outputBuffer: Int = 16
+  ): ZStream[R, E, A] =
+    ZStream {
+      for {
+        output  <- Queue.bounded[stream.Take[E, A]](outputBuffer).toManaged(_.shutdown)
+        runtime <- ZIO.runtime[R].toManaged_
+        maybeStream <- ZManaged.effectTotal {
+                         register(k => runtime.unsafeRunToFuture(stream.Take.fromPull(k).flatMap(output.offer)))
```
Comment (iravid) [inline streams/js/src/main/scala/zio/stream/platform.scala:114]: Do we want to handle interruption from `output.offer` here too?

File: streams/js/src/main/scala/zio/stream/platform.scala (hunk 1 above)
Comment (luis3m) [reply streams/js/src/main/scala/zio/stream/platform.scala]: Done

File: streams/js/src/main/scala/zio/stream/platform.scala (hunk 3 above)
Comment (luis3m) [reply streams/js/src/main/scala/zio/stream/platform.scala]: Done

File: streams/js/src/main/scala/zio/stream/platform.scala (hunk 4 above)
Comment (luis3m) [reply streams/js/src/main/scala/zio/stream/platform.scala]: Done

File: streams/js/src/main/scala/zio/stream/platform.scala (hunk 5 above)
Comment (luis3m) [reply streams/js/src/main/scala/zio/stream/platform.scala]: Done

File: streams/js/src/main/scala/zio/stream/platform.scala (hunk 6 above)
Comment (luis3m) [reply streams/js/src/main/scala/zio/stream/platform.scala]: Done

File: streams/js/src/main/scala/zio/stream/platform.scala (hunk 7 above)
Comment (luis3m) [reply streams/js/src/main/scala/zio/stream/platform.scala]: Good catch, updated.

File: streams/js/src/main/scala/zio/stream/platform.scala (hunk 2 above)
Comment (luis3m) [reply streams/js/src/main/scala/zio/stream/platform.scala]: Done

**Review by iravid — APPROVED** [review]
Great work @luis3m 🙏

**Pull request merged**